### PR TITLE
Generate some PropertyDeclaration methods by hand 🐉🐲

### PR DESCRIPTION
**Pull request by nox**
We rely on https://github.com/rust-lang/rfcs/pull/2195 to make some enums share the same discriminants by definition.

This fixes https://bugzilla.mozilla.org/show_bug.cgi?id=1428285.

This change is [Reviewable](https://reviewable.io/reviews/servo/servo/19956)

### CODE REVIEW

**Comment by nox** [comment]
https://treeherder.mozilla.org/#/jobs?repo=try&revision=0cccd012152be2f205e31bb00254900e85215152

File: components/style/properties/properties.mako.rs [hunk 1]
```diff
@@ -1438,34 +1441,126 @@ impl PropertyId {
...
+            "doc": "A custom property declaration.",
+        },
+    ]
+%>
+
 /// Servo's representation for a property declaration.
 #[cfg_attr(feature = "gecko", derive(MallocSizeOf))]
-#[derive(Clone, PartialEq)]
+#[derive(PartialEq)]
+#[repr(u8)]
```
Comment (SimonSapin) [inline components/style/properties/properties.mako.rs]: Doesn’t this have more than 256 variants?

Comment (nox) [reply components/style/properties/properties.mako.rs]: Yes. :D Just realised with the geckotry failing to build.

**Comment by nox** [comment]
I made the enum be `#[repr(u16)]`.

https://treeherder.mozilla.org/#/jobs?repo=try&revision=680ae344ebda08a71b94be64049039d2328abd33

**Comment by nox** [comment]
```
[nox:~/src/gecko] 2s 130
⌀ size XUL.old
__TEXT	__DATA	__OBJC	others	dec	hex
88387584	5271552	0	65503232	159162368	97ca000

[nox:~/src/gecko]
⌀ ../bloaty/build/bloaty -d symbols -n 100000000 XUL.old | grep -w PropertyDeclaration
   0.0%  45.1Ki style::properties::PropertyDeclaration::to_css::h70aa9127e631f236                 45.1Ki   0.0%
   0.0%  32.8Ki _$LT$style..properties..PropertyDeclaration$u20$as$u20$core..cmp..PartialEq$GT$:  32.8Ki   0.0%
   0.0%  32.6Ki __ZN79_$LT$style..properties..PropertyDeclaration$u20$as$u20$core..cmp..PartialE  32.6Ki   0.0%
   0.0%  29.7Ki __ZN77_$LT$style..properties..PropertyDeclaration$u20$as$u20$core..clone..Clone$  29.7Ki   0.0%
   0.0%  29.3Ki __ZN77_$LT$style..properties..PropertyDeclaration$u20$as$u20$core..clone..Clone$  29.3Ki   0.0%
   0.0%  27.9Ki __ZN77_$LT$style..properties..PropertyDeclaration$u20$as$u20$core..clone..Clone$  27.9Ki   0.0%
   0.0%  27.6Ki _$LT$style..properties..PropertyDeclaration$u20$as$u20$core..clone..Clone$GT$::c  27.6Ki   0.0%
   0.0%  15.7Ki __ZN87_$LT$style..properties..PropertyDeclaration$u20$as$u20$malloc_size_of..Mal  15.7Ki   0.0%
   0.0%  9.78Ki style::properties::PropertyDeclaration::parse_into::h35ee1deaf54c46b6             9.78Ki   0.0%
   0.0%  4.17Ki style::properties::PropertyDeclaration::id::h9f691ba9ad1fcbbc                     4.17Ki   0.0%
   0.0%     320 style::properties::PropertyDeclaration::get_system::h4a3d3afe5a217d86                320   0.0%
   0.0%      32 style::properties::PropertyDeclaration::get_css_wide_keyword::he3ca52066fa1abf7       32   0.0%

[nox:~/src/gecko] 2s
⌀ size XUL.new
__TEXT	__DATA	__OBJC	others	dec	hex
88190976	5271552	0	65482752	158945280	9795000

[nox:~/src/gecko]
⌀ ../bloaty/build/bloaty -d symbols -n 100000000 XUL.new | grep -w PropertyDeclaration
   0.0%  45.2Ki style::properties::PropertyDeclaration::to_css::hc9cb9df128aea03f                 45.2Ki   0.0%
   0.0%  22.8Ki _$LT$style..properties..PropertyDeclaration$u20$as$u20$core..cmp..PartialEq$GT$:  22.8Ki   0.0%
   0.0%  22.3Ki _$LT$style..properties..PropertyDeclaration$u20$as$u20$core..clone..Clone$GT$::c  22.3Ki   0.0%
   0.0%  15.6Ki __ZN87_$LT$style..properties..PropertyDeclaration$u20$as$u20$malloc_size_of..Mal  15.6Ki   0.0%
   0.0%  9.89Ki style::properties::PropertyDeclaration::parse_into::ha8f87489fe4f89f8             9.89Ki   0.0%
   0.0%     320 style::properties::PropertyDeclaration::get_system::h07f9efc836477132                320   0.0%
   0.0%      96 style::properties::PropertyDeclaration::id::hd928f19c20f0e1bc                         96   0.0%
   0.0%      32 style::properties::PropertyDeclaration::get_css_wide_keyword::hfd74e2964361369d       32   0.0%
```

It looks good but I'm not sure I should trust those results.

Cc @bholley @emilio 

Also, I've yet to make a geckotry with the latest commits I added.

https://treeherder.mozilla.org/#/jobs?repo=try&revision=d0b009ecd44b0a7327c507b37a9b188e2a5b7e7b

**Comment by nox** [comment]
Silly me had not put `#[inline]` on the `Clone` and `PartialEq` methods, that explains why there was only one `Clone` in `XUL.new`:

```
⌀ ../bloaty/build/bloaty -d symbols -n 1000000000 XUL.new | grep -w PropertyDeclaration
   0.0%  45.2Ki style::properties::PropertyDeclaration::to_css::hc9cb9df128aea03f                 45.2Ki   0.0%
   0.0%  23.0Ki _$LT$style..properties..PropertyDeclaration$u20$as$u20$core..cmp..PartialEq$GT$:  23.0Ki   0.0%
   0.0%  22.7Ki __ZN77_$LT$style..properties..PropertyDeclaration$u20$as$u20$core..clone..Clone$  22.7Ki   0.0%
   0.0%  22.7Ki __ZN79_$LT$style..properties..PropertyDeclaration$u20$as$u20$core..cmp..PartialE  22.7Ki   0.0%
   0.0%  22.3Ki __ZN77_$LT$style..properties..PropertyDeclaration$u20$as$u20$core..clone..Clone$  22.3Ki   0.0%
   0.0%  22.2Ki __ZN77_$LT$style..properties..PropertyDeclaration$u20$as$u20$core..clone..Clone$  22.2Ki   0.0%
   0.0%  21.8Ki _$LT$style..properties..PropertyDeclaration$u20$as$u20$core..clone..Clone$GT$::c  21.8Ki   0.0%
   0.0%  15.6Ki __ZN87_$LT$style..properties..PropertyDeclaration$u20$as$u20$malloc_size_of..Mal  15.6Ki   0.0%
   0.0%  9.89Ki style::properties::PropertyDeclaration::parse_into::ha8f87489fe4f89f8             9.89Ki   0.0%
   0.0%     320 style::properties::PropertyDeclaration::get_system::h07f9efc836477132                320   0.0%
   0.0%      96 style::properties::PropertyDeclaration::id::hd928f19c20f0e1bc                         96   0.0%
   0.0%      32 style::properties::PropertyDeclaration::get_css_wide_keyword::hfd74e2964361369d       32   0.0%
```

**Comment by nox** [comment]
Everything is orange. Will investigate tomorrow. Probably some dumb error somewhere.

**Comment by nox** [comment]
Latest results with bloaty, and the missing `#[inline]` on `clone` and `eq`:

```
[nox:~/src/gecko]
⌀ size XUL.old
__TEXT	__DATA	__OBJC	others	dec	hex
88387584	5271552	0	65503232	159162368	97ca000

[nox:~/src/gecko]
⌀ ../bloaty/build/bloaty -d symbols -n 1000000000 XUL.old | grep -w PropertyDeclaration
   0.0%  45.1Ki style::properties::PropertyDeclaration::to_css::h70aa9127e631f236                 45.1Ki   0.0%
   0.0%  32.8Ki _$LT$style..properties..PropertyDeclaration$u20$as$u20$core..cmp..PartialEq$GT$:  32.8Ki   0.0%
   0.0%  32.6Ki __ZN79_$LT$style..properties..PropertyDeclaration$u20$as$u20$core..cmp..PartialE  32.6Ki   0.0%
   0.0%  29.7Ki __ZN77_$LT$style..properties..PropertyDeclaration$u20$as$u20$core..clone..Clone$  29.7Ki   0.0%
   0.0%  29.3Ki __ZN77_$LT$style..properties..PropertyDeclaration$u20$as$u20$core..clone..Clone$  29.3Ki   0.0%
   0.0%  27.9Ki __ZN77_$LT$style..properties..PropertyDeclaration$u20$as$u20$core..clone..Clone$  27.9Ki   0.0%
   0.0%  27.6Ki _$LT$style..properties..PropertyDeclaration$u20$as$u20$core..clone..Clone$GT$::c  27.6Ki   0.0%
   0.0%  15.7Ki __ZN87_$LT$style..properties..PropertyDeclaration$u20$as$u20$malloc_size_of..Mal  15.7Ki   0.0%
   0.0%  9.78Ki style::properties::PropertyDeclaration::parse_into::h35ee1deaf54c46b6             9.78Ki   0.0%
   0.0%  4.17Ki style::properties::PropertyDeclaration::id::h9f691ba9ad1fcbbc                     4.17Ki   0.0%
   0.0%     320 style::properties::PropertyDeclaration::get_system::h4a3d3afe5a217d86                320   0.0%
   0.0%      32 style::properties::PropertyDeclaration::get_css_wide_keyword::he3ca52066fa1abf7       32   0.0%

[nox:~/src/gecko] 3s
⌀ size XUL.new
__TEXT	__DATA	__OBJC	others	dec	hex
88317952	5271552	0	65495040	159084544	97b7000

[nox:~/src/gecko]
⌀ ../bloaty/build/bloaty -d symbols -n 1000000000 XUL.new | grep -w PropertyDeclaration
   0.0%  45.2Ki style::properties::PropertyDeclaration::to_css::he42a36e94b3deab2                 45.2Ki   0.0%
   0.0%  23.0Ki _$LT$style..properties..PropertyDeclaration$u20$as$u20$core..cmp..PartialEq$GT$:  23.0Ki   0.0%
   0.0%  22.7Ki __ZN77_$LT$style..properties..PropertyDeclaration$u20$as$u20$core..clone..Clone$  22.7Ki   0.0%
   0.0%  22.7Ki __ZN79_$LT$style..properties..PropertyDeclaration$u20$as$u20$core..cmp..PartialE  22.7Ki   0.0%
   0.0%  22.3Ki __ZN77_$LT$style..properties..PropertyDeclaration$u20$as$u20$core..clone..Clone$  22.3Ki   0.0%
   0.0%  22.2Ki __ZN77_$LT$style..properties..PropertyDeclaration$u20$as$u20$core..clone..Clone$  22.2Ki   0.0%
   0.0%  21.8Ki _$LT$style..properties..PropertyDeclaration$u20$as$u20$core..clone..Clone$GT$::c  21.8Ki   0.0%
   0.0%  15.6Ki __ZN87_$LT$style..properties..PropertyDeclaration$u20$as$u20$malloc_size_of..Mal  15.6Ki   0.0%
   0.0%  9.89Ki style::properties::PropertyDeclaration::parse_into::ha8f87489fe4f89f8             9.89Ki   0.0%
   0.0%     320 style::properties::PropertyDeclaration::get_system::h07f9efc836477132                320   0.0%
   0.0%      96 style::properties::PropertyDeclaration::id::hd928f19c20f0e1bc                         96   0.0%
   0.0%      32 style::properties::PropertyDeclaration::get_css_wide_keyword::hfd74e2964361369d       32   0.0%
```

Geckotry seems green: https://treeherder.mozilla.org/#/jobs?repo=try&revision=2cb22ae5e6be97fcc968dc965827234d54f26b47&selectedJob=160528329

File: components/style/properties/helpers.mako.rs [hunk 2]
```diff
@@ -903,6 +903,7 @@
     <%call expr="longhand(name,
                           predefined_type=length_type,
                           logical=logical,
+                          is_size_type=True,
```
Comment (emilio) [inline components/style/properties/helpers.mako.rs]: Let's rename this to something that screams out more, like `is_gecko_size_type_hack` or something.

File: components/style/properties/properties.mako.rs [hunk 3]
```diff
@@ -1441,28 +1442,94 @@ impl PropertyId {
...
         if property.predefined_type and not property.is_vector and not property.is_size_type:
             ty = "::values::specified::{}".format(property.predefined_type)
         else:
             ty = "longhands::{}::SpecifiedValue".format(property.ident)
         if property.boxed:
             ty = "Box<{}>".format(ty)
-    %>
-    /// ${property.name}
-    ${property.camel_case}(${ty}),
+        variants.append({
```
Comment (emilio) [inline components/style/properties/properties.mako.rs]: Can you document why do we need this complexity here, please?

File: components/style/properties/properties.mako.rs [hunk 4]
```diff
@@ -1623,23 +1646,9 @@ impl PropertyDeclaration {
...
-                debug_assert!(false, "unreachable");
-                // This value is never used, but having an expression of the same "shape"
-                // as for other variants helps the optimizer compile this `match` expression
-                // to a lookup table.
-                LonghandId::BackgroundColor
-            }
-        };
-        PropertyDeclarationId::Longhand(longhand_id)
+        PropertyDeclarationId::Longhand(
+            unsafe { *(self as *const _ as *const LonghandId) },
```
Comment (emilio) [inline components/style/properties/properties.mako.rs]: Can we debug-assert this somehow at least, to ensure nobody messes up? I guess if they mess up it'd be completely orange but...

Also, please add a comment about why this is ok here so people don't scream :).

File: components/style/properties/properties.mako.rs [hunk 5]
```diff
@@ -1572,6 +1599,28 @@ pub struct VariableDeclaration {
...
+impl ToCss for VariableDeclaration {
+    fn to_css<W>(&self, dest: &mut CssWriter<W>) -> fmt::Result
+    where
+        W: fmt::Write,
+    {
+        // https://drafts.csswg.org/css-variables/#variables-in-shorthands
+        match self.value.from_shorthand {
+            // Normally, we shouldn't be printing variables here if they came from
+            // shorthands. But we should allow properties that came from shorthand
+            // aliases. That also matches with the Gecko behavior.
```
Comment (emilio) [inline components/style/properties/properties.mako.rs]: Can you mention here that this is just a hack for `-moz-transform`, which is really a shorthand for who knows what reason?

Also we should try to unship that, so feel free to also add a `FIXME(emilio)` over there so I eventually do it.

File: components/style/properties/properties.mako.rs [hunk 6]
```diff
@@ -314,6 +313,32 @@ impl Clone for PropertyDeclaration {
...
+                &*(other as *const _ as *const PropertyDeclarationVariantRepr<()>);
+            if this_repr.tag != other_repr.tag {
+                return false;
+            }
+            match *self {
+                % for ty, variants in groups.iteritems():
+                ${" | ".join("{}(ref this)".format(v) for v in variants)} => {
+                    let other_repr =
+                        &*(other as *const _ as *const PropertyDeclarationVariantRepr<${ty}>);
+                    *this == other_repr.value
```
Comment (emilio) [inline components/style/properties/properties.mako.rs]: Wonderfully evil :(

File: components/style/properties/properties.mako.rs [hunk 7]
```diff
@@ -267,7 +268,6 @@ pub mod animated_properties {
 %>
 
 /// Servo's representation for a property declaration.
-#[cfg_attr(feature = "gecko", derive(MallocSizeOf))]
 #[repr(u16)]
 pub enum PropertyDeclaration {
```
Comment (emilio) [inline components/style/properties/properties.mako.rs]: Can you add a general overview of why don't we derive stuff here and similar?

**Review by emilio — APPROVED** [review]
r=me, with the comments and a different name for `is_size_type` (feel free to rename it on a followup commit if it's easier).

**Comment by emilio** [comment]
If @SimonSapin sanity-checked this too it'd be nice, since he had concerns about doing the same kind of optimization in https://bugzilla.mozilla.org/show_bug.cgi?id=1428285#c2, it seems.

**Comment by SimonSapin** [comment]
My concern in that comment was about transmuting the return type of `std::mem::discriminant` to u64. This function was very deliberately designed to not return a plain integer, but instead return an opaque type that makes few guarantees.

This PR however relies on memory layout guarantees documented by https://github.com/rust-lang/rfcs/pull/2195 when `#[repr(u16)]` is used on an enum type.

@nox Please add a link to https://github.com/rust-lang/rfcs/pull/2195 and a short explanation to the PR message.

**Comment by nox** [comment]
I added a comment in the commit message.

@bors-servo r=emilio,SimonSapin

**Comment by ionutgoldan** [comment]
These changes landed some fine performance improvements! Thank you!

== Change summary for alert #11504 (as of Tue, 06 Feb 2018 12:06:36 GMT) ==

Improvements:

  5%  Stylo Servo_DeclarationBlock_SetPropertyById_WithInitialSpace_Bench osx-10-10 opt      549,031.92 -> 519,840.33
  2%  Stylo Servo_DeclarationBlock_GetPropertyById_Bench windows7-32 opt                     231,626.04 -> 226,308.08

For up to date results, see: https://treeherder.mozilla.org/perf.html#/alerts?id=11504

**Comment by bholley** [comment]
There are some more wins in #19966 that might also be attributable to this.

Awesome job @nox!

**Pull request merged**